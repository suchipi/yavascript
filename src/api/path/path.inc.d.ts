/** An object that represents a filesystem path. */
declare class Path {
  /** The character used to separate path segments on this OS. */
  static readonly OS_SEGMENT_SEPARATOR: "/" | "\\";

  /**
   * The character used to separate entries within the PATH environment
   * variable on this OS.
   */
  static readonly OS_ENV_VAR_SEPARATOR: ":" | ";";

  /**
   * A list of suffixes that could appear in the filename for a program on the
   * current OS. For instance, on Windows, programs often end with ".exe".
   *
   * On Unix-like OSes, this is empty, On Windows, it's based on `env.PATHEXT`.
   */
  static readonly OS_PROGRAM_EXTENSIONS: ReadonlySet<string>;

  /** Split one or more path strings into an array of path segments. */
  static splitToSegments(inputParts: Array<string> | string): Array<string>;

  /**
   * Search the provided path string or strings for a path separator character,
   * and return it. If none is found, return `fallback`, which defaults to the
   * OS's path segment separator.
   */
  static detectSeparator<Fallback extends string | null = string>(
    input: Array<string> | string,
    // @ts-ignore might be instantiated with a different subtype
    fallback: Fallback = Path.OS_SEGMENT_SEPARATOR
  ): string | Fallback;

  /** Join together one or more paths. */
  static join(...inputs: Array<string | Path | Array<string | Path>>): Path;

  /**
   * Turns the input path(s) into an absolute path by resolving all `.` and `..`
   * segments, using `pwd()` as a base dir to use when resolving leading `.` or
   * `..` segments.
   */
  static resolve(...inputs: Array<string | Path | Array<string | Path>>): Path;

  /**
   * Concatenates the input path(s) and then resolves all non-leading `.` and
   * `..` segments.
   */
  static normalize(
    ...inputs: Array<string | Path | Array<string | Path>>
  ): Path;

  /**
   * Return whether the provided path is absolute; that is, whether it
   * starts with either `/` or a drive letter (ie `C:`).
   */
  static isAbsolute(path: string | Path): boolean;

  /**
   * An array of the path segments that make up this path.
   *
   * For `/tmp/foo.txt`, it'd be `["", "tmp", "foo.txt"]`.
   *
   * For `C:\something\somewhere.txt`, it'd be `["C:", "something", "somewhere.txt"]`.
   */
  segments: Array<string>;

  /**
   * The path separator that should be used to turn this path into a string.
   *
   * Will be either `/` or `\`.
   */
  separator: string;

  /** Create a new Path object using the provided input(s). */
  constructor(...inputs: Array<string | Path | Array<string | Path>>);

  /** Create a new Path object using the provided segments and separator. */
  static from(segments: Array<string>, separator: string): Path;

  /**
   * Create an absolute path by `concat`ting `subpaths` onto this Path (which is
   * presumed to be an absolute path) and then using `normalize()` on the
   * result. If the result is not an absolute path, an error will be thrown.
   */
  resolve(...subpaths: Array<string | Path>): Path;

  /**
   * Resolve all non-leading `.` and `..` segments in this path.
   */
  normalize(): Path;

  /**
   * Create a new Path by appending additional path segments onto the end of
   * this Path's segments.
   *
   * The returned path will use this path's separator.
   */
  concat(...other: Array<string | Path | Array<string | Path>>): Path;

  /**
   * Return whether this path is absolute; that is, whether it starts with
   * either `/`, `\`, or a drive letter (ie `C:`).
   */
  isAbsolute(): boolean;

  /**
   * Make a second Path object containing the same segments and separator as
   * this one.
   */
  clone(): this;

  /**
   * Express this path relative to `dir`.
   *
   * @param dir - The directory to create a new path relative to.
   * @param options - Options that affect the resulting path.
   */
  relativeTo(
    dir: Path | string,
    options?: {
      /**
       * Defaults to false. When true, a leading `./` will be omitted from the
       * path, if present. Note that a leading `../` will never be omitted.
       */
      noLeadingDot?: boolean;
    }
  ): Path;

  /**
   * Turn this path into a string by joining its segments using its separator.
   */
  toString(): string;

  /**
   * Alias for `toString`; causes Path objects to be serialized as strings when
   * they (or an object referencing them) are passed into JSON.stringify.
   */
  toJSON(): string;

  /**
   * Return the final path segment of this path. If this path has no path
   * segments, the empty string is returned.
   */
  basename(): string;

  /**
   * Return the trailing extension of this path. The `options` parameter works
   * the same as the global `extname`'s `options` parameter.
   */
  extname(options?: { full?: boolean }): string;
}
