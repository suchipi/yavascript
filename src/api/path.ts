import * as os from "os";
import { makeErrorWithProperties } from "../error-with-properties";

function validateSegments(segments: Array<string>): Array<string> {
  return segments.filter((part, index) => {
    // first part can be "" to represent left side of root "/"
    if (index === 0) return true;
    return Boolean(part);
  });
}

export class Path {
  static OS_SEGMENT_SEPARATOR = os.platform === "win32" ? "\\" : "/";
  static OS_ENV_VAR_SEPARATOR = os.platform === "win32" ? ";" : ":";

  static splitToSegments(inputParts: Array<string> | string): Array<string> {
    if (!Array.isArray(inputParts)) {
      inputParts = [inputParts];
    }

    return validateSegments(
      inputParts.map((part) => part.split(/(?:\/|\\)+/g)).flat(1)
    );
  }

  static detectSeparator<Fallback extends string | null = string>(
    input: Array<string> | string,
    // @ts-ignore might be instantiated with a different subtype
    fallback: Fallback = Path.OS_SEGMENT_SEPARATOR
  ): string | Fallback {
    let testStr = input;
    if (Array.isArray(input)) {
      testStr = input.join("|");
    }

    for (const char of testStr) {
      if (char === "/") {
        return "/";
      } else if (char === "\\") {
        return "\\";
      }
    }

    return fallback;
  }

  static join(...inputs: Array<string | Path | Array<string | Path>>): string {
    return new Path(...inputs).toString();
  }

  static resolve(
    ...inputs: Array<string | Path | Array<string | Path>>
  ): string {
    return new Path(...inputs).resolve().toString();
  }

  static normalize(
    ...inputs: Array<string | Path | Array<string | Path>>
  ): string {
    return new Path(...inputs).normalize().toString();
  }

  static isAbsolute(path: string): boolean {
    return new Path(path).isAbsolute();
  }

  static tag(
    _strings: TemplateStringsArray,
    ..._values: ReadonlyArray<string | Path | Array<string | Path>>
  ) {
    const strings = [..._strings];
    const values = [..._values];

    const zipped: Array<string | Path | Array<string | Path>> = [];

    let sourceIsStrings = true;
    while (strings.length + values.length > 0) {
      const next = sourceIsStrings ? strings.shift() : values.shift();
      if (next != null) {
        zipped.push(next);
      }
      sourceIsStrings = !sourceIsStrings;
    }

    if (zipped[0] === "") {
      zipped.shift();
    }

    return new Path(...zipped);
  }

  static tagUsingBase(dir: string | Path): typeof Path.tag {
    const ret = (strings, ...values) => {
      const stringsClone = Object.assign([...strings], {
        raw: [...strings.raw],
      });
      stringsClone[0] = Path.join(dir, strings[0]);
      stringsClone.raw[0] = Path.join(dir, strings.raw[0]);
      return Path.tag(stringsClone, ...values);
    };
    Object.defineProperty(ret, "name", {
      value: `tagUsingBase(${dir})`,
      configurable: true,
    });
    return ret;
  }

  segments: Array<string>;
  separator: string;

  constructor(...inputs: Array<string | Path | Array<string | Path>>) {
    const parts = inputs
      .flat(1)
      .map((part) => (typeof part === "string" ? part : part.segments))
      .flat(1);

    this.segments = Path.splitToSegments(parts);
    this.separator = Path.detectSeparator(parts);
  }

  static from(
    segments: Array<string>,
    separator: string = Path.OS_SEGMENT_SEPARATOR
  ) {
    const path = new Path();
    path.segments = validateSegments(segments);
    path.separator = separator;
    return path;
  }

  resolve(from: string | Path = os.getcwd()): Path {
    if (this.isAbsolute()) {
      return this.normalize();
    }

    const fromPath = typeof from === "string" ? new Path(from) : from;
    const result = fromPath.concat(this).normalize();
    if (!result.isAbsolute()) {
      throw makeErrorWithProperties(
        `Could not resolve ${this.toString()} from ${fromPath.toString()}`,
        {
          this: this.toString(),
          from: fromPath.toString(),
        }
      );
    }
    return result;
  }

  normalize(): Path {
    // we clone this cause we're gonna mutate it
    const segments = [...this.segments];

    const newSegments: Array<string> = [];
    let currentSegment: string | undefined;
    while (segments.length > 0) {
      currentSegment = segments.shift();

      switch (currentSegment) {
        case ".": {
          if (newSegments.length === 0) {
            newSegments.push(currentSegment);
          }
          break;
        }
        case "..": {
          if (newSegments.length === 0) {
            newSegments.push(currentSegment);
          } else {
            newSegments.pop();
          }

          break;
        }
        default: {
          if (currentSegment != null) {
            newSegments.push(currentSegment);
          }
          break;
        }
      }
    }

    return Path.from(newSegments, this.separator);
  }

  concat(other: string | Path | Array<string | Path>): Path {
    const otherSegments = new Path(other).segments;
    return Path.from(this.segments.concat(otherSegments), this.separator);
  }

  isAbsolute(): boolean {
    const firstPart = this.segments[0];

    // empty first component indicates that path starts with leading slash
    if (firstPart === "") return true;

    // windows drive
    if (/^[A-Za-z]:/.test(firstPart)) return true;

    // TODO: windows UNC paths (not supported very well by these path APIs at all, yet)

    return false;
  }

  toString() {
    const result = this.segments.join(this.separator);
    if (result == "") {
      return "/";
    } else {
      return result;
    }
  }
}
