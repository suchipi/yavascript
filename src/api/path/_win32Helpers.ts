import * as os from "quickjs:os";

const WIN32_DRIVE_LETTER_REGEXP = /^[A-Za-z]:$/;

export function appendSlashIfWindowsDriveLetter(pathString: string): string {
  if (os.platform === "win32" && WIN32_DRIVE_LETTER_REGEXP.test(pathString)) {
    return pathString + Path.OS_SEGMENT_SEPARATOR;
  } else {
    return pathString;
  }
}
