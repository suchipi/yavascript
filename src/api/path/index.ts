import { Path } from "./path";
import "./help-text/apply-help-text";
export { Path };
