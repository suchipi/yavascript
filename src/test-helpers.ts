import { spawn, Options as SpawnOptions } from "first-base";
import child_process from "child_process";
import path from "path";
import print from "@suchipi/print";
import * as inspectOptions from "./inspect-options";

export const TMP = child_process
  .execSync("realpath /tmp", { encoding: "utf-8" })
  .trim();

export const rootDir = (...parts: Array<string>) =>
  path.resolve(__dirname, "..", ...parts);

function getBinaryPath(platform: string, arch: string) {
  if (platform === "win32") {
    return path.resolve(__dirname, "..", "bin", "windows", "yavascript.exe");
  } else if (platform === "darwin") {
    if (arch.startsWith("arm")) {
      return path.resolve(__dirname, "..", "bin", "darwin-arm", "yavascript");
    } else {
      return path.resolve(__dirname, "..", "bin", "darwin", "yavascript");
    }
  } else if (platform === "linux") {
    return path.resolve(__dirname, "..", "bin", "linux", "yavascript");
  } else {
    throw new Error("Unsupported platform: " + platform);
  }
}

var binaryPath = getBinaryPath(
  process.platform,
  process.env.YS_TESTS_ARCH || process.arch
);

export { binaryPath, getBinaryPath };

export type EvaluateResult = {
  stdout: string;
  stderr: string;
  code: number | null;
  error: boolean;
};

export async function evaluate(
  code: string,
  options: SpawnOptions = {}
): Promise<EvaluateResult> {
  const runContext = spawn(binaryPath, ["-e", code], options);
  await runContext.completion;
  return runContext.result;
}

export function inspect(value: any): string {
  // options to inspect here match what is given to console.log
  return print(value, inspectOptions.forPrint);
}

export function cleanOutput(input: string): string {
  return input
    .replace(/yavascript:\d+/g, "yavascript")
    .replace(/\/bin\/[^/]+\/yavascript/g, "/bin/.../yavascript")
    .replace(new RegExp(rootDir(), "g"), "<rootDir>");
}

export function cleanResult(input: EvaluateResult): EvaluateResult {
  return {
    ...input,
    stdout: cleanOutput(input.stdout),
    stderr: cleanOutput(input.stderr),
  };
}
