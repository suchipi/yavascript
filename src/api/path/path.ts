import * as os from "quickjs:os";
import { assert } from "../assert";
import { types } from "../types";
import { is } from "../is";
import { makeErrorWithProperties } from "../../error-with-properties";
import { appendSlashIfWindowsDriveLetter } from "./_win32Helpers";

function validateSegments(
  segments: Array<string>,
  separator: string
): Array<string> {
  return segments.filter((part, index) => {
    // first part can be "" to represent left side of root "/"
    // second part can be "" to support windows UNC paths
    if (part === "" && index === 0) {
      return true;
    } else if (
      part === "" &&
      index === 1 &&
      separator === "\\" &&
      segments[0] === ""
    ) {
      return true;
    }

    return Boolean(part);
  });
}

class Path {
  static OS_SEGMENT_SEPARATOR = os.platform === "win32" ? "\\" : "/";
  static OS_ENV_VAR_SEPARATOR = os.platform === "win32" ? ";" : ":";

  static splitToSegments(inputParts: Array<string> | string): Array<string> {
    assert.type(
      inputParts,
      types.or(types.string, types.arrayOf(types.string))
    );

    if (!Array.isArray(inputParts)) {
      inputParts = [inputParts];
    }

    const separator = Path.detectSeparator(inputParts);

    return validateSegments(
      inputParts.map((part) => part.split(/(?:\/|\\)/g)).flat(1),
      separator
    );
  }

  static detectSeparator<Fallback extends string | null = string>(
    input: Array<string> | string,
    // @ts-ignore might be instantiated with a different subtype
    fallback: Fallback = Path.OS_SEGMENT_SEPARATOR
  ): string | Fallback {
    assert.type(input, types.or(types.string, types.arrayOf(types.string)));
    assert.type(fallback, types.or(types.string, types.null));

    let testStr = input;
    if (Array.isArray(input)) {
      testStr = input.join("|");
    }

    for (const char of testStr) {
      if (char === "/") {
        return "/";
      } else if (char === "\\") {
        return "\\";
      }
    }

    return fallback;
  }

  static join(...inputs: Array<string | Path | Array<string | Path>>): string {
    assert.type(
      inputs,
      types.arrayOf(
        types.or(
          types.string,
          types.Path,
          types.arrayOf(types.or(types.string, types.Path))
        )
      )
    );

    return new Path(...inputs).toString();
  }

  static resolve(
    ...inputs: Array<string | Path | Array<string | Path>>
  ): string {
    assert.type(
      inputs,
      types.arrayOf(
        types.or(
          types.string,
          types.Path,
          types.arrayOf(types.or(types.string, types.Path))
        )
      )
    );

    return new Path(...inputs).resolve().toString();
  }

  static normalize(
    ...inputs: Array<string | Path | Array<string | Path>>
  ): string {
    assert.type(
      inputs,
      types.arrayOf(
        types.or(
          types.string,
          types.Path,
          types.arrayOf(types.or(types.string, types.Path))
        )
      )
    );

    return new Path(...inputs).normalize().toString();
  }

  static isAbsolute(path: string | Path): boolean {
    assert.type(path, types.or(types.string, types.Path));

    if (is(path, types.Path)) {
      return path.isAbsolute();
    } else {
      return new Path(path).isAbsolute();
    }
  }

  static tag(
    _strings: TemplateStringsArray,
    ..._values: ReadonlyArray<string | Path | Array<string | Path>>
  ) {
    assert.type(_strings, types.arrayOf(types.string));
    assert.type(
      _values,
      types.arrayOf(
        types.or(
          types.string,
          types.Path,
          types.arrayOf(types.or(types.string, types.Path))
        )
      )
    );

    // clone both arrays so we can `.shift()` from them without worry of
    // mutating the original inputs
    const strings = [..._strings];
    const values = [..._values];

    const zipped: Array<string | Path | Array<string | Path>> = [];

    let sourceIsStrings = true;
    while (strings.length + values.length > 0) {
      const next = sourceIsStrings ? strings.shift() : values.shift();
      if (next != null) {
        zipped.push(next);
      }
      sourceIsStrings = !sourceIsStrings;
    }

    if (zipped[0] === "") {
      zipped.shift();
    }

    return new Path(...zipped);
  }

  static tagUsingBase(dir: string | Path): typeof Path.tag {
    assert.type(dir, types.or(types.string, types.Path));

    const ret = (strings, ...values) => {
      const stringsClone = Object.assign([...strings], {
        raw: [...strings.raw],
      });
      stringsClone[0] = Path.join(dir, strings[0]);
      stringsClone.raw[0] = Path.join(dir, strings.raw[0]);
      return Path.tag(stringsClone, ...values);
    };
    Object.defineProperty(ret, "name", {
      value: `tagUsingBase(${dir})`,
      configurable: true,
    });
    return ret;
  }

  segments: Array<string>;
  separator: string;

  constructor(...inputs: Array<string | Path | Array<string | Path>>) {
    assert.type(
      inputs,
      types.arrayOf(
        types.or(
          types.string,
          types.Path,
          types.arrayOf(types.or(types.string, types.Path))
        )
      )
    );

    const parts = inputs
      .flat(1)
      .map((part) => (typeof part === "string" ? part : part.segments))
      .flat(1);

    this.segments = Path.splitToSegments(parts);
    this.separator = Path.detectSeparator(parts);
  }

  static from(
    segments: Array<string>,
    separator: string = Path.OS_SEGMENT_SEPARATOR
  ) {
    assert.type(segments, types.arrayOf(types.string));
    assert.type(separator, types.string);

    const path = new Path();
    path.segments = validateSegments(segments, separator);
    path.separator = separator;
    return path;
  }

  resolve(from: string | Path = os.getcwd()): Path {
    assert.type(from, types.or(types.string, types.Path));

    if (this.isAbsolute()) {
      return this.normalize();
    }

    const fromPath = typeof from === "string" ? new Path(from) : from;
    const result = fromPath.concat(this).normalize();
    if (!result.isAbsolute()) {
      throw makeErrorWithProperties(
        `Could not resolve ${this.toString()} from ${fromPath.toString()}`,
        {
          this: this.toString(),
          from: fromPath.toString(),
        }
      );
    }
    return result;
  }

  normalize(): Path {
    // we clone this cause we're gonna mutate it
    const segments = [...this.segments];

    const newSegments: Array<string> = [];
    let currentSegment: string | undefined;
    while (segments.length > 0) {
      currentSegment = segments.shift();

      switch (currentSegment) {
        case ".": {
          if (newSegments.length === 0) {
            newSegments.push(currentSegment);
          }
          break;
        }
        case "..": {
          if (newSegments.length === 0) {
            newSegments.push(currentSegment);
          } else {
            newSegments.pop();
          }

          break;
        }
        default: {
          if (currentSegment != null) {
            newSegments.push(currentSegment);
          }
          break;
        }
      }
    }

    return Path.from(newSegments, this.separator);
  }

  concat(other: string | Path | Array<string | Path>): Path {
    assert.type(
      other,
      types.or(
        types.string,
        types.Path,
        types.arrayOf(types.or(types.string, types.Path))
      )
    );

    const otherSegments = new Path(other).segments;
    return Path.from(this.segments.concat(otherSegments), this.separator);
  }

  isAbsolute(): boolean {
    const firstPart = this.segments[0];

    // empty first component indicates that path starts with leading slash.
    // could be unix fs root, or windows unc path
    if (firstPart === "") return true;

    // windows drive
    if (/^[A-Za-z]:/.test(firstPart)) return true;

    return false;
  }

  clone() {
    return (this.constructor as typeof Path).from(
      this.segments,
      this.separator
    );
  }

  relativeTo(
    dir: Path | string,
    options: { noLeadingDot?: boolean } = {}
  ): Path {
    if (!is(dir, types.Path)) {
      dir = new Path(dir);
    }

    const ownSegments = [...this.segments];
    const dirSegments = [...dir.segments];

    while (ownSegments[0] === dirSegments[0]) {
      ownSegments.shift();
      dirSegments.shift();
    }

    if (dirSegments.length === 0) {
      if (options.noLeadingDot) {
        return Path.from(ownSegments, this.separator);
      } else {
        return Path.from([".", ...ownSegments], this.separator);
      }
    } else {
      const dotDots = dirSegments.map((_) => "..");
      return Path.from([...dotDots, ...ownSegments]);
    }
  }

  toString() {
    let result = this.segments.join(this.separator);
    if (result == "") {
      return "/";
    } else {
      result = appendSlashIfWindowsDriveLetter(result);
      return result;
    }
  }

  toJSON() {
    return this.toString();
  }
}

// .toString() needs to return a value starting with "class" for pheno.coerce
// to see this function as a class, and therefore coerce it into the type
// "instanceOf(Path)" when it appears in a function that accepts a type
// validator (ie. `is` or `assert.type`).
//
// Normally, `Path.toString()` would already start with "class", because we
// defined it using class syntax. But, as part of compiling yavascript, the
// source code is converted to QuickJS bytecode, and that bytecode
// representation does not preserve Function bodies, so `Path.toString()`
// changes. It instead returns "function Path() {\n    [native code]\n}".
//
// Explicitly overriding it like this ensures that it has the correct value
// even after bytecode conversion.
Object.defineProperty(Path, "toString", {
  value: () => "class Path {\n    [native code]\n}",
});

export { Path };
