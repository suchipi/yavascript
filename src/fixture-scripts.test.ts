import fs from "fs";
import path from "path";
import { spawn } from "first-base";
import { binaryPath, cleanResult } from "./test-helpers";

const scriptsDir = path.resolve(__dirname, "test_fixtures", "scripts");

const scripts = fs.readdirSync(scriptsDir);
for (const script of scripts) {
  test(script, async () => {
    const run = spawn(binaryPath, [script], { cwd: scriptsDir });
    await run.completion;
    expect(cleanResult(run.result)).toMatchSnapshot();
  });
}
