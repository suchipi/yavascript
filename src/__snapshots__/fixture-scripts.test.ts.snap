// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__filename-and-__dirname.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  __filename: "<rootDir>/src/test_fixtures/scripts/__filename-and-__dirname.js"
  __dirname: "<rootDir>/src/test_fixtures/scripts"
}
{
  __filename: "<rootDir>/src/test_fixtures/scripts/__filename-and-__dirname-2.js"
  __dirname: "<rootDir>/src/test_fixtures/scripts"
}
",
}
`;

exports[`__filename-and-__dirname-2.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "",
}
`;

exports[`http-import-error.js 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: Failed to load module (httpStatusCode = 404, url = "http://suchipi.com/there-is-nothing-at-this-path")
  at somewhere
{
  httpStatusCode: 404
  url: "http://suchipi.com/there-is-nothing-at-this-path"
  httpResponseHeaders: <redacted>
  httpResponseBody: <redacted>
}
",
  "stdout": "",
}
`;

exports[`https-import-error.js 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: Failed to load module (httpStatusCode = 404, url = "https://suchipi.com/there-is-nothing-at-this-path")
  at somewhere
{
  httpStatusCode: 404
  url: "https://suchipi.com/there-is-nothing-at-this-path"
  httpResponseHeaders: <redacted>
  httpResponseBody: <redacted>
}
",
  "stdout": "",
}
`;

exports[`importing-from-npm-protocol.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  Null prototype
  Sealed
  
  Component: Function "b" {}
  Fragment: Function "k" {}
  cloneElement: Function "F" {}
  createContext: Function "G" {}
  createElement: Function "y" {}
  createRef: Function "_" {}
  default: null
  h: -> {root}.createElement
  hydrate: Function "E" {}
  isValidElement: Function "t" {}
  options: {
    __e: Function "__e" {}
  }
  render: Function "D" {}
  toChildArray: Function "S" {}
}
",
}
`;

exports[`importing-json.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  hello: "there"
}
",
}
`;

exports[`npm-proto-import-error.js 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: Failed to load module (httpStatusCode = 404, url = "https://cdn.skypack.dev/@suchipi/this-package-does-not-exist")
  at somewhere
{
  httpStatusCode: 404
  url: "https://cdn.skypack.dev/@suchipi/this-package-does-not-exist"
  httpResponseHeaders: <redacted>
  httpResponseBody: <redacted>
}
",
  "stdout": "",
}
`;

exports[`some-coffeescript.coffee 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "object
object
false
",
}
`;

exports[`some-json.json 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "",
}
`;

exports[`some-jsx.jsx 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: "a"
  props: null
  key: null
}
true
false
true
false
false
",
}
`;

exports[`some-tsx.tsx 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: "a"
  props: null
  key: null
}
true
false
true
false
false
",
}
`;

exports[`some-typescript.ts 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "undefined
object true
path: <rootDir>/src/test_fixtures/scripts
<rootDir>/src/test_fixtures/scripts
",
}
`;

exports[`toplevel-await.coffee 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "4 Function "open" {}
",
}
`;

exports[`toplevel-await.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "4 Function "open" {}
",
}
`;

exports[`toplevel-await.ts 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: nope!
  at somewhere
",
  "stdout": "",
}
`;

exports[`toplevel-await.tsx 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: "a"
  props: null
  key: null
}
Function "open" {}
",
}
`;

exports[`toplevel-await-and-export-together.js 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: You cannot use toplevel await and export in the same file. Either remove the toplevel await (by wrapping stuff with an immediately-invoked async function) or remove the exports.
  at somewhere
{
  filename: "<rootDir>/src/test_fixtures/scripts/toplevel-await-and-export-together.js"
}
",
  "stdout": "",
}
`;

exports[`type-coercion-via-assert.ts 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "no error message from hi String
Expected value of type string, but received null
Expected value of type string, but received 24
no error message from hi hi
Expected value of type exactString("hi"), but received "yes"
Expected value of type number, but received "something"
no error message from 42 Number
",
}
`;

exports[`type-coercion-via-is.ts 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "true
false
false
true
false
false
true
",
}
`;
